versioning/worker-versions.md

```diff
@@ -0,0 +1,191 @@
+workflows which break history compatability (but not interface compatability),
+or would like to safely roll out expected-to-be-compatible changes.
+
+At a high level:
+* Provide a simple methodology for versioning entire workers (this proposal)
+* Allow clients to point at task queue `foo`, and automatically route new workflow executions
+  to the latest version of the `foo` queue.
+* Provide good notifications and information about when old workers may be
+  decommissioned
+* Ensure histories replayed on older versions fail fast with informative errors
+* Easy visibility into what versions deployed workers are using
+
+Building on this base set of features, we plan to eventually implement
+automatic safe-rollout of changes, and possibly tools like k8s operators
+which can automatically manage worker deployments.
+
+## Safe-rollout
+
+Intentionally avoiding deep into details here, the foundation laid here should enable
+us, as well as users, to more easily implement automated and safe rollout
+of changes to workers. A process like this one can become the de-facto experience:
+
+* If deploying a known-breaking change, new workflows are only processed by
+  the new version, existing open workflows are processed by old workers.
+* If deploying an expected-to-be-compatible change, new workflow _tasks_ are
+  attempted by new workers, and if they succeed, that workflow will subsequently
+  only be processed by new workers. If they fail, execution goes back to old
+  workers.
+
+# Using opaque whole-worker-identifiers as worker versions
+
+Thanks to @jlegrone for his great feedback encouraging us to revive this approach.
+
+We already have a field called `binaryChecksum` (or similar) in `WorkerOptions`
+today in SDKs. This field's intent is to uniquely identify a specific worker
+build. Since we believe versioning entire workers is the easiest solution, we
+can lean on this existing field.
+
+Deprecating the name `binaryChecksum`and replacing it with something more 
+accurate like `workerBuildId` would be appropriate, since most of our SDKs
+don't even produce a binary.
+
+With this approach, users who wish to make use of the improved versioning 
+features should set the `workerBuildId` during their CI process to a value that
+is unique for that worker's implementation. _Implementation_ rather than 
+_build_ is an important distinction here, since rebuilds of identical
+code should (ideally) produce workers wit the same ID. It is not necessary to
+do this, and it is varyingly difficult depending on the language, but when
+done it enables users to produce workers with the same ID from their CI builds
+if they didn't touch worker code, but something else (ex: The CI definition itself).
+This involves some work on the part of the user, but is more reliable than expecting
+worker authors to increment a version number.
+
+Unfortunately it is infeasible to automatically set an appropriate ID for many languages.
```

Comment:
Go 1.18 has support for it and it automatically emmits git SHA to every binary. And git SHA is very good candidate for build Id.

Reply:
> And git SHA is very good candidate for build Id.

Not necessarily, especially if the process is used for more than just worker handling. This is not to be confused with what might be good for "worker binary ID" because _this_ ID will actually affect which workflows are scheduled where and for many they may want multiple workers with different build IDs running in the same process (even for the same task queue). Also, you don't necessarily want to change a worker build ID/version on every commit, especially if you don't link the versions together. Imagine having to either have code linking every commit together, or having to keep a process around for every commit just so it can finish workflows started on it.

I am not sure we should assume that binary checksum is a good default value for this as the beginning of this section says. But if we expose this with an easy way for them to set it, we don't really care then and versioning is opt-in anyways. They can choose how unique they want to be.

Reply:
I don't mean to say we should lean on _the way it is set by default_. I mean to say we can lean on _the fact that it already exists_. This is also why I want to deprecate that old name and rename it. The checksum by itself is effectively useless. It was really always a stand in for "this worker's implementation" and the name conflated the approach Go used by default with the goal of the field.

We certainly should encourage users to set it as part of their CI process. Git SHA is I believe a pretty reasonable thing to use, because it succinctly expresses the intent ("the set of code we used to build the worker"). It is true that for cases like multiple workers, you'll want to use something else (but that could easily be `<sha>-workera` and `<sha>-workerb`.

And the idea here is you really _do_ want to change it on every commit. You're going to lean on the mechanisms we'll build to make sure each worker deployment can take workflows without breaking them. Every commit is not linked together, but every deployment is - by virtue of the fact that it's the next (assumed incompatible) default version, unless you explicitly specify your intent (via API call) that this newly deployed version should be compatible with some extant one.

If you don't want this behavior, just don't opt in to the feature and go wild west like the existing system, but the whole intent here is to have a safe-by-default approach for people who want it (of which we know of many)

Reply:
Is a company that will use the Git SHA expected to either: 1) add links on every deployment to the other Git SHA if compatible (manually or via some ugly automated thing), and/or 2) keep a worker around for every Git SHA that may have started a workflow?

I would think most companies would want semver or a more specific way to signal compatibility. They may make hundreds of commits/deployments without a single backwards incompatibility, there's no need to tie the commit to the worker version.

Of course they have the freedom to do what they want, but when we get to documenting suggested approaches, I think a human-approved/changeable version number makes the most sense since they have to make the decision as a human on whether it's compatible anyways.